### PR TITLE
Improve brew functionality (WIP)

### PR DESCRIPTION
**Pull request by norman-abramovitz**
1) added version compare function for Tyler to look at
2) Added ability for brew to deal with multiple versions installed
3) Added ability to warn user a brew upgrade exists
4) Used log functions in place of echo commands in some places

### CODE REVIEW

File: bin/update_toc [hunk 1]
```diff
@@ -11,12 +11,111 @@ err() { log "[err] $*" >&2 ; }
...
+            [[ "${_fill}" == "_ver1" ]] && _ver1[i]=0 || _ver2[i]=0 
+        done
+    fi
+
+    for ((i=0; i<${_maxlen}; i++))
+    do
+        (( 10#${_ver1[i]} > 10#${_ver2[i]} )) && return 1
+        (( 10#${_ver1[i]} < 10#${_ver2[i]} )) && return 2
+    done
+    return 0
```
Comment (7hunderbird) [inline bin/update_toc]: Will this always return 0 or will it return 0 only if we have no breaks before that point?

File: bin/update_toc [hunk 2]
```diff
@@ -11,12 +11,111 @@ err() { log "[err] $*" >&2 ; }
...
+
+# version number compare - = returns  0; < returns 1; > returns 22
+
+vercomp () {
+
+    [[ "${1}" == "${2}" ]] && return 0
+
+    local i _ver1 _ver2
+    local _low="" _fill _maxlen
+    _ver1=($(IFS=.; set -- $1; for a in "${@}"; do echo ${a:-0}; done))
```
Comment (7hunderbird) [inline bin/update_toc]: Can we get an example of the "before" and "after" in a comment of what the in-field separator is doing to the `_ver1` variable(s), etc. ?

**Comment by 7hunderbird** [comment]
Separating the `check_install` into a another function called `check_install_with_brew` is breaking the API.

A `check_install`, checks if it needs to be installed, if it's not there or the right version, install it based on the prescribed way.

I'd prefer we keep the API consistent for all things we check, be that a brew package or a node module.  And merge where you're going with the `check_install_with_brew` back into the `check_install` function.

File: bin/update_toc (hunk 1 above)
Comment (norman-abramovitz) [inline bin/update_toc]: No,  I will either return 0, 1 or 2.   The return statement exits the vercmp function.  The return statement only works inside functions.

File: bin/update_toc (hunk 2 above)
Comment (norman-abramovitz) [inline bin/update_toc]: sure,   I will update the code.   For your edification now, the code is expecting with version string inputs of u.v.w.x.y.z.   The number of components does not matter and a version string like 1...2 gets converted to 1.0.0.2.

**Comment by 7hunderbird** [comment]
@norman-abramovitz thanks for all the hard work on this.  A practice I used before was to add (WIP) to a PR until it was ready for review.  Then add "RFC @7hunderbird" when the Request For Change was ready.  If this sounds good for you, let's try that method for working through a work-in-progress pull request.

**Comment by norman-abramovitz** [comment]
works for me.  thanks for the tip.

**Comment by jhunt** [comment]
You can also push to origin/BRANCH without making a PR.

**Comment by norman-abramovitz** [comment]
I create the PR because I thought what I did initially would be acceptable.   I am redoing things based upon Tyler's input.   I think I will close the PR until I am finished.

**Pull request closed without merging**